docs/rfcs/1005-realm-key-rotation.md

```diff
@@ -0,0 +1,821 @@
+file content a search on the workspace mountpoint).
+In addition since Parsec 2.16 the manifests are eagerly downloaded (instead of the
+previous lazy approach), so the reencryption really has no security benefit.
+
+And now the final blow: the maintenance preventing read/write access was such a strong
+limitation that we ended up adding a feature to keep read access possible.
+Now what that means is the client has to in fact deal with up to *two* keys:
+
+1. when a maintenance starts, the client receives the new key
+2. it overwrites the realm key with the new one it the user manifest
+3. now it tries to access some realm data
+   - the maintenance is done, so the new key works. All is good.
+   - the maintenance is still ongoing, so the client has to look into the previous
+     versions of the user manifest to find the old key. This is messy :'(
+
+## 2 General approach
+
+The idea is as follow:
+
+- remove the whole maintenance logic: a realm is always accessible in read&write
+- remove the whole re-encryption logic: realm vlobs are never reencrypted
+- remove the whole message logic: we use a special ad-hoc command to access the
+  realm's keys (see below)
+- a realm now has it manifests encrypted with multiple keys, at any time a new
+  key can be added
+- each key for a given realm has an index
+- only the last key should be used to encrypt new manifest
+- all those keys are serialized together in a `RealmKeysBundle` which is encrypted by the
+  "keys-bundle key" symmetric key and stored in a dedicate table in the server database
+- the "keys-bundle key" is serialized in a `RealmKeysBundleAccess`. This document is signed
+  by the author of the key rotation and encrypted for each member of the realm with their
+  public key
+- for sharing a realm, the last `RealmKeysBundleAccess` is encrypted for the new user.
+  The `RealmKeysBundleAccess` also contains the workspace label.
+- keys-bundle & keys-bundle access are never stored on-disk client-side: they are instead
+  lazily fetched from server when needed and kept on RAM.
+- realm initial key is lazily created by doing the very first key rotation (typically during
+  minimal sync right after having created the realm).
+- key rotation also involves the creation of a `RealmKeyRotationCertificate` certificate to
+  allow auditability (knowing when and by whom the key rotation occurred)
+- key rotation certificate contains the index of the key along with it algorithm and a
+  canary (i.e. an encrypted payload to ensure the key obtained from the keys-bundle is genuin)
+- key rotation consists of creating a new `RealmKeysBundle` with one additional
+  key, then submit it (along with the `RealmKeyRotationCertificate` and the per-user
+  `RealmKeysBundleAccess`) to the server with the new `workspace_rotate_keys` API command.
+- given each key rotation is expected to add a single key to the bundle, the key index is
+  also used to identify the key rotation and the keys-bundle
```

Comment:
This smells a little bit... using the index to identify key rotation and keys-bundle imposes a strong constraint. Are we sure there can never be "holes"? I don't have a counter-example yet, but I'm thinking in some data corruption or alike...

What about using hashes for identification + tracking the last hash (rotation)? They can even be chained by having pointers to the "parent" rotation (yes, a git-ish solution ^^). It is certainly more complex, but it might be worth it to at least compare it to the index-based solution.

Reply:
> What about using hashes for identification

using a hash of a key is potentially worrying given it can leak info if not done properly

hence the only other solution is using a uuid to identify a key, but given we want key and key-rotation-certificate to be related 1-to-1 (for simplification), then it's more logical to identify them by the order they are created

Reply:
> Can a KeyBundle be added/created for another reason besides a key rotation? (Maintenance, Security, Data suppression...)

for simplification we don't want that !

> If we consider data suppression, there might be some cases where keys are no longer needed

In theory yes, in practice it's most likely there is still some data that use any given key (think of the README in a repo that haven't been updated in ages, also in Parsec we handle History so old data must still be accessible).
In the end it's no use bothering with cleanup up old keys.